hhnk_threedi_tools/qgis/layer_structure.py

```diff
@@ -0,0 +1,401 @@
+
+    def get_layers_from_df(self) -> pd.Series:
+        """Get a list of all layers"""
+        layers = {}
+        for index, row in self.df_full.iterrows():
+            layer = self._get_layer_from_row(row)
+            layers[layer.id] = layer
+        return pd.Series(layers)
+
+    def get_themes_from_df(self) -> pd.Series:
+        """Get a series with themes.
+        First read the layers with self.get_layers_from_df"""
+        theme_col_names = [i for i in self.df_full.keys() if i.startswith("theme_")]
+        themes = {}
+        for theme_col_name in theme_col_names:
+            theme_filter = self.df_full[theme_col_name] == True
+
+            layer_ids = self.df_full.loc[theme_filter, "layer"].apply(lambda x: x.id).tolist()
+            name = theme_col_name[6:]  # remove str theme_
+            themes[name] = QgisThemeSettings(name=name, layer_ids=layer_ids)
+        return pd.Series(themes)
+
+    def _get_layer_from_row(self, row) -> QgisLayerSettings:
+        """Evalualte df row and create a QgisLayer instance"""
+
+        def eval_filenames(name, suffix, folder, revisions):
+            """some filenames are variables"""
+            if "." in name:
+                return eval(name)
+            else:
+                return f"{name}.{suffix}"
+
+        def eval_qml(qmldir, qmlnames, HHNK_THREEDI_PLUGIN_DIR):
+            """create list of qmlpaths with row.qmldir and row.qmlnames"""
+            qmldir = eval(qmldir)
+            if qmlnames.startswith("["):
+                # Already a list
+                qmlnames = eval(qmlnames)
+            else:  # Set single qml name to list
+                qmlnames = [qmlnames]
+            qmlpaths = []
+            for name in qmlnames:
+                qmlpaths.append(hrt.Folder(qmldir).full_path(name))
+            return qmlpaths
+
+        # required for eval
+        folder = self.folder
+        revisions = self.revisions
+        HHNK_THREEDI_PLUGIN_DIR = self.plugin_dir
+
+        # Voor wms staat de volledige link die nodig is in row.wms_source.
+        file = None
+        if row.filename:
+            file = hrt.Folder(eval(str(row.filedir))).full_path(
+                eval_filenames(row.filename, row.suffix, folder=folder, revisions=self.revisions)
+            )
+
+        group_lst = []
+        if row.group_lst:
+            group_lst = eval(row.group_lst)
+
+        qml_lst = []
+        if row.qmlnames:
+            qml_lst = eval_qml(
+                qmldir=row.qmldir, qmlnames=row.qmlnames, HHNK_THREEDI_PLUGIN_DIR=HHNK_THREEDI_PLUGIN_DIR
+            )
+
+        l = QgisLayerSettings(
+            name=row.qgis_name,
+            file=file,
+            filters=row.filters,
+            wms_source=row.wms_source,
+            qml_lst=qml_lst,
+            group_lst=group_lst,
+            load_layer=row.subject in self.subjects,
+            subject=row.subject,
+        )
+        return l
+
+    def run(self):
```

Comment:
Why a run-method? Suggestion:
- run self.run() at __init__, or
- convert LayerStructure to a @dataclass and put all code from `def run(self):` to `def __post_init__(self):`